### PR TITLE
Manage plugin lifetime

### PR DESCRIPTION
**Pull request by rageagainsthepc**

### CODE REVIEW

**Comment by rageagainsthepc** [comment]
* TODO: Update plugin README.md

File: plugins/apitracing/src/lib/ApiTracing.cpp [hunk 1]
```diff
@@ -90,29 +69,18 @@ namespace ApiTracing
...
-        return true;
+    void ApiTracing::unload()
+    {
+        tracer->removeHooks();
     }
 }
+
+std::unique_ptr<IPlugin>
+VmiCore::Plugin::init(PluginInterface* pluginInterface,
+                      std::shared_ptr<IPluginConfig> config, // NOLINT(performance-unnecessary-value-param)
```
Comment (K-Mayer) [inline plugins/apitracing/src/lib/ApiTracing.cpp:82]: nit: I don't think the // NOLINT is needed, right?

File: plugins/inmemoryscanner/src/lib/Scanner.cpp [hunk 2]
```diff
@@ -13,20 +13,23 @@ using VmiCore::Plugin::PluginInterface;
...
-          inMemResultsLogger(this->pluginInterface->newNamedLogger(INMEMORY_LOGGER_NAME))
+          logger(pluginInterface->newNamedLogger(INMEMORY_LOGGER_NAME)),
+          inMemResultsLogger(pluginInterface->newNamedLogger(INMEMORY_LOGGER_NAME))
     {
         logger->bind({{VmiCore::WRITE_TO_FILE_TAG, LOG_FILENAME}});
         inMemResultsLogger->bind(
             {{VmiCore::WRITE_TO_FILE_TAG, (this->configuration->getOutputPath() / TEXT_RESULT_FILENAME).string()}});
+
+        pluginInterface->registerProcessTerminationEvent([this](auto&& a)
+                                                         { scanProcess(std::forward<decltype(a)>(a)); });
```
Comment (K-Mayer) [inline plugins/inmemoryscanner/src/lib/Scanner.cpp:32]: Why is this different to the usage in ApiTracing? Both callbacks should take the same parameter type, right?

File: plugins/template/src/lib/Template.cpp [hunk 3]
```diff
@@ -6,48 +6,28 @@
 #define BUILD_VERSION "TestBuild"
 #endif
 
+#include "Template.h"
```
Comment (K-Mayer) [inline plugins/template/src/lib/Template.cpp:9]: nit: Include order

File: vmicore/src/lib/plugins/PluginSystem.cpp [hunk 4]
```diff
@@ -190,50 +181,47 @@ namespace VmiCore
...
+            throw PluginException(pluginName,
+                                  fmt::format("Plugin API Version {} is incompatible with VmiCore API Version {}",
+                                              *pluginApiVersion,
+                                              Plugin::PluginInterface::API_VERSION));
         }
 
-        auto pluginInitFunction = reinterpret_cast<init_f>(dlsym(libraryHandle, "init"));
+        auto pluginInitFunction = reinterpret_cast<decltype(Plugin::init)*>(
+            dlsym(libraryHandle,
+                  "_ZN7VmiCore6Plugin4initEPNS0_15PluginInterfaceENSt3__110shared_ptrINS0_13IPluginConfigEEENS3_"
```
Comment (K-Mayer) [inline vmicore/src/lib/plugins/PluginSystem.cpp:214]: wat? oO

File: vmicore/src/lib/VmiHub.cpp [hunk 5]
```diff
@@ -177,24 +177,33 @@ namespace VmiCore
...
-        setupSignalHandling();
-        waitForEvents();
+            eventStream->sendReadyEvent();
+
+            setupSignalHandling();
+            waitForEvents();
+        }
+        catch (const PluginException& e)
+        {
+            logger->warning("Failed to initialize plugin", {{"Plugin", e.plugin()}, {"Exception", e.what()}});
```
Comment (K-Mayer) [inline vmicore/src/lib/VmiHub.cpp:200]: This should be an error because it ends the program run.

File: vmicore/src/lib/plugins/PluginSystem.cpp [hunk 6]
```diff
@@ -254,14 +242,26 @@ namespace VmiCore
...
     }
 
-    void PluginSystem::passShutdownEventToRegisteredPlugins()
+    void PluginSystem::unloadPlugins()
     {
         vmiInterface->flushV2PCache(LibvmiInterface::flushAllPTs);
         vmiInterface->flushPageCache();
 
-        for (auto& shutdownCallback : registeredShutdownCallbacks)
+        for (auto& plugin : plugins)
```
Comment (K-Mayer) [inline vmicore/src/lib/plugins/PluginSystem.cpp:250]: I think I would move both plugin-init and unloading into the PluginSystem. In the current version you have the init-loop over all plugins in VmiHup and the unload-loop in PluginSystem. Feels a bit off.

File: vmicore/src/lib/plugins/PluginSystem.cpp [hunk 7]
```diff
@@ -254,14 +242,26 @@ namespace VmiCore
...
         {
-            shutdownCallback();
+            try
+            {
+                plugin.second->unload();
+            }
+            catch (const std::exception& e)
+            {
+                logger->warning("Error occurred while unloading plugin",
```
Comment (K-Mayer) [inline vmicore/src/lib/plugins/PluginSystem.cpp:258]: I think this also should be an error.

File: plugins/template/src/lib/Template.cpp (hunk 3 above)
Comment (rageagainsthepc) [reply plugins/template/src/lib/Template.cpp:9]: What do you mean? Includes are sorted automatically.

**Review by cakeless — APPROVED** [review]

**Pull request merged**